File: search/binary-search/binary-search.js
function search(arr, value) {
  let min = 0;
  let max = arr.length - 1;

  while (min <= max) {
    let middle = Math.floor((max + min) / 2);
    if (arr[middle] === value) return middle;
    if (arr[middle] > value) max = middle - 1;
    else min = middle + 1;
  }
  return -1;
}

console.log(search([2, 3, 4, 5, 6, 123, 989, 1000], 1000) === 7);
console.log(search([1, 2, 3, 6, 7, 10, 12], 3) === 2);
console.log(search([5, 32, 43, 57], 570) === -1);

File: search/binary-search/binary-search.cpp
#include <iostream>


int binarySearch(int arr[], int length, int val) {
    int left = 0; 
    int right = length - 1;
    int middle; 
    while(left <= right) {
        middle = (right + left) / 2;
        if (arr[middle] == val) return middle;
        else if (arr[middle] > val) right = middle - 1;
        else left = middle + 1;
    }

    return -1;
}

int main() {
    std::cout << "BINARY SEARCH!" << std::endl;
    int arr[10] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9}; 
    
    for (int i = 0; i < 10; i++) {
        std::cout << (binarySearch(arr, sizeof(arr)/ sizeof(int), i) == i ? "Test passed" : "Test failed") << std::endl;
    }

    // test for the value that doesn't exist in the array. Should return true
    std::cout << (binarySearch(arr, sizeof(arr)/ sizeof(int), 10) == -1 ? "Test passed" : "Test failed") << std::endl;
}